### ISSUE TITLE
Either vs Try catch

### ISSUE THREAD

**Issue by jabrena** [jabrena/typed-errors#68]
In Java, the use of the `Either` type versus traditional `try-catch` blocks presents several advantages, particularly when dealing with functional programming paradigms. Here are the key advantages of using `Either` over `try-catch`:

1. **Explicit Handling of Success and Failure**:
   - **`Either`**: Encapsulates both success and failure paths in the type system. An `Either<L, R>` can be either a `Left<L>` (typically used for failure) or a `Right<R>` (typically used for success), making it clear that a function can return either a success or an error.
   - **`try-catch`**: The error path is implicit and relies on exceptions, which are not part of the method signature.

2. **Type Safety**:
   - **`Either`**: Errors and successes are typed, ensuring that the possible error types are known at compile time. This reduces the chance of unhandled exceptions and makes the code more robust.
   - **`try-catch`**: Exceptions are not part of the method signature, so the types of exceptions that might be thrown are not immediately apparent without inspecting the method's implementation or documentation.

3. **Functional Composition**:
   - **`Either`**: Can be easily composed and chained using functional methods like `map`, `flatMap`, etc. This makes it easier to construct pipelines of operations that can handle errors gracefully.
   - **`try-catch`**: Handling exceptions in chained methods often requires nested try-catch blocks, which can become cumbersome and harder to read.

4. **Immutability and Side-effect Management**:
   - **`Either`**: Emphasizes immutability and encourages functions without side effects, aligning with functional programming principles.
   - **`try-catch`**: Exception handling can lead to side effects and mutable state changes, which can be harder to track and reason about.

5. **Better Error Handling and Recovery**:
   - **`Either`**: Allows for more granular and controlled error handling. Since errors are part of the return type, handling them becomes a conscious part of the code flow.
   - **`try-catch`**: Errors can be caught and handled at different levels in the call stack, which might lead to less predictable and more error-prone error recovery.

6. **Improved Readability and Maintainability**:
   - **`Either`**: Code using `Either` can be more declarative and often more concise, which can improve readability and maintainability.
   - **`try-catch`**: Can lead to verbose and scattered error-handling code, making it harder to follow the main logic of the program.

7. **Encouragement of Pure Functions**:
   - **`Either`**: Encourages the writing of pure functions (functions without side effects), which can lead to more predictable and testable code.
   - **`try-catch`**: Exception handling within methods can lead to side effects, making the code less predictable.

Here’s a simple example to illustrate the difference:

**Using `Either`**:
```java
import io.vavr.control.Either;

public Either<Error, Integer> parseInt(String str) {
    try {
        return Either.right(Integer.parseInt(str));
    } catch (NumberFormatException e) {
        return Either.left(new Error("Invalid number format"));
    }
}

// Usage
Either<Error, Integer> result = parseInt("123");
result.map(value -> value * 2)
      .getOrElseThrow(error -> new RuntimeException(error.getMessage()));
```

**Using `try-catch`**:
```java
public Integer parseInt(String str) throws NumberFormatException {
    return Integer.parseInt(str);
}

// Usage
try {
    Integer result = parseInt("123");
    result *= 2;
} catch (NumberFormatException e) {
    throw new RuntimeException("Invalid number format", e);
}
```

In summary, using `Either` can lead to more explicit, type-safe, and maintainable code by making both success and error cases part of the function's signature and promoting functional programming practices.

**Comment by jabrena** [comment]
- https://g.co/gemini/share/da1927a0094d

**Comment by jabrena** [comment]
M- Here's a brief comparison of Either and Try/Catch blocks for error handling in Scala:
Try/Catch
 * Purpose: Handles unexpected exceptions during program execution.
 * Returns: Either a successful result or a Throwable representing the exception.
 * Use cases: Ideal for catching unforeseen exceptions and recovering from errors gracefully.
Either
 * Purpose: Represents two possible outcomes: success or failure.
 * Returns: Either a successful value or an error value of a specific type.
 * Use cases: Suitable for signaling expected errors with specific information for handling.
Choosing between Either and Try/Catch
 * Use Try/Catch for exception handling when the error type is unknown or unexpected.
 * Use Either for signaling expected errors with clear error types for specific handling.
